### PR TITLE
Add basic IPv6 support

### PR DESCRIPTION
**Pull request by LolloneS**
Add basic IPv6 support.

The ROBOT test works fine for standard cases (e.g. it performs an IPv6 test when provided with an IPv6 address), but fails in corner cases (e.g. IPv6 address given, but the service only accepts IPv4 connections). 

The default IP version is IPv4: this means that when checking example.com, if it is possible to resolve it to an IPv4 address, that will be the used one. In other words, in order to use IPv6 you can either pass an IPv6 address as host directly, or have a service that is *only* exposed on IPv6 and therefore does only resolve to it. I never seen a case such as the latter before, so I strongly suggest just passing the IPv6 address as host if you want to have it checked. Otherwise, the default behavior will be the standard IPv4 one.

@bbc2

### CODE REVIEW

**Comment by bbc2** [comment]
I can't understand the diff because they are mixed with formatting changes.  Could you separate the two?  Actually, upstream probably won't want to reformat all their code so it's probably better to not touch the rest of their code and try to use the same style.

File: robot-detect [hunk 1]
```diff
@@ -186,19 +206,37 @@ enable_fastopen = os.path.exists("/proc/sys/net/ipv4/tcp_fastopen")
 
 try:
     ip = socket.gethostbyname(args.host)
+    server = (ip, args.port)
```
Comment (bbc2) [inline robot-detect]: Why keep the use of `socket.gethostbyname`?  Wouldn't it make sense to use `socket.getaddrinfo` in all cases, since it works the same way for IPv4 and IPv6?

Additionally, I think that it would enable the script to handle domain names for any type of IP address.

File: robot-detect [hunk 2]
```diff
@@ -34,20 +34,40 @@ MSG_FASTOPEN = 0x20000000
...
+            s.connect(server)
+            cert_raw = s.getpeercert(binary_form=True)
+            cert_dec = x509.load_der_x509_certificate(cert_raw, default_backend())
+            s.close()
+            return (
+                cert_dec.public_key().public_numbers().n,
+                cert_dec.public_key().public_numbers().e,
+            )
+        except OSError as e:
+            print("Error: %s \nYou are probably trying to connect to an IPv6 service via IPv4 or vice-versa. "
```
Comment (bbc2) [inline robot-detect]: Actually, this error could be due to a different issue, right?  It might be better to keep it generic.  The user should figure out why they couldn't connect.

Also, an IP family mismatch should be very rare if the first address from `socket.getaddrinfo` is used, because it should return the IPv4 first for IPv4-only clients.

File: robot-detect [hunk 3]
```diff
@@ -34,20 +34,40 @@ MSG_FASTOPEN = 0x20000000
...
 
 
-def get_rsa_from_server(server, port):
+def create_socket(ip_family=4, timeout=None):
+    if ip_family == 4:
+        s = socket.socket()
+        if timeout is not None:
+            s.settimeout(timeout)
+        return s
+    s = socket.socket(socket.AF_INET6)
```
Comment (bbc2) [inline robot-detect]: Why is a different kind of socket required?

**Review by bbc2 — CHANGES_REQUESTED** [review]

File: robot-detect (hunk 1 above)
Comment (LolloneS) [reply robot-detect]: Yes, I can also use that. The idea is that `gethostbyname` does not support IPv6, and therefore we only need to do the `getaddrinfo` part if there is no IPv4 host to check in the first place. However, I can simply switch to `getaddrinfo` and use the first (?) result. The main goal was to avoid looking for IPv6 in the first place if we have an IPv4 service.

File: robot-detect (hunk 2 above)
Comment (LolloneS) [reply robot-detect]: Yes it could. From my tests it seems like some other exception is thrown before if there are other problems, so in this case I don't see many other possibilities for the error to appear if not for the reported one. Still, I can be more generic.

Yes, the mismatch is very rare for that reason.

File: robot-detect (hunk 3 above)
Comment (LolloneS) [reply robot-detect]: `AF_INET` seems to be IPv4-only.

From https://docs.python.org/3/library/socket.html#socket-families, on `AF_INET`:

> This behavior is not compatible with IPv6, therefore, you may want to avoid these if you intend to support IPv6 with your Python programs.

File: robot-detect (hunk 1 above)
Comment (bbc2) [reply robot-detect]: Yes.  Actually it would probably make more sense to create the socket with `socket.create_connection`, so that you don't have to deal with the list returned by `getaddrinfo`.

Comment (LolloneS) [reply robot-detect]: Yes, I can definitely use that and it eases things as far as creating the socket is concerned, but we still need the IP at least to put it into the CSV not breaking compatibility with previous versions. Whatever I do, I guess I still need to get the IP either using `gethostbyname` (IPv4) or `getaddrinfo` (both, but more convoluted). I would take only the part I need from the current commit and save the IP somewhere, what do you think about it?

Comment (bbc2) [reply robot-detect]: Could you use `socket.getpeername()` or `socket.getsockname()` to get the information and pass it to other sockets?

File: robot-detect (hunk 3 above)
Comment (bbc2) [reply robot-detect]: I see.  But I'm not sure we need that here.  You could pass the parameters that you compute the very first time with `getaddrinfo` or `create_connection`/`getpeername`/`getsockname` to that function, and then pass them further to `socket.socket` and `socket.connect`.  This way you don't need this extra `if-then-else`.

File: robot-detect (hunk 1 above)
Comment (LolloneS) [reply robot-detect]: Done, got it to be much cleaner.

File: robot-detect (hunk 3 above)
Comment (LolloneS) [reply robot-detect]: Unfortunately, in the `oracle` function I need to create a `socket` but not a connection, in order to allow the `fastopen` if available. However I think I have managed to solve this issue in a polished way (line 193). In this case it is simply not possible to bypass the choice of IP type, since the `sendto` function used in case `fastopen` is available should not work on an open connection.

File: robot-detect [hunk 4]
```diff
@@ -184,21 +182,32 @@ else:
 # We only enable TCP fast open if the Linux proc interface exists
 enable_fastopen = os.path.exists("/proc/sys/net/ipv4/tcp_fastopen")
 
+server = (args.host, args.port)
+
 try:
-    ip = socket.gethostbyname(args.host)
+    connection = socket.create_connection(server, timeout=timeout)
+    ip = connection.getpeername()[0]
+    ip_family = socket.AF_INET if type(ipaddress.ip_address(ip)) == ipaddress.IPv4Address else socket.AF_INET6
```
Comment (bbc2) [inline robot-detect]: Probably easier to check `ip_address.version == 4`

**Review by bbc2 — CHANGES_REQUESTED** [review]
Looks good!  Just one small comment.
